typespec-extension/src/code-model-builder.ts

```diff
@@ -1302,119 +1305,138 @@ export class CodeModelBuilder {
     });
     op.addParameter(parameter);
 
+    const jsonMergePatch = operationIsJsonMergePatch(sdkHttpOperation);
+
+    const schemaIsPublicBeforeProcess =
+      schema instanceof ObjectSchema && (schema as SchemaUsage).usage?.includes(SchemaContext.Public);
+
     this.trackSchemaUsage(schema, { usage: [SchemaContext.Input] });
 
     if (op.convenienceApi) {
       // model/schema does not need to be Public or Internal, if it is not to be used in convenience API
       this.trackSchemaUsage(schema, { usage: [op.internalApi ? SchemaContext.Internal : SchemaContext.Public] });
     }
 
-    if (operationIsJsonMergePatch(sdkHttpOperation)) {
+    if (jsonMergePatch) {
       this.trackSchemaUsage(schema, { usage: [SchemaContext.JsonMergePatch] });
     }
     if (op.convenienceApi && operationIsMultipart(sdkHttpOperation)) {
       this.trackSchemaUsage(schema, { serializationFormats: [KnownMediaType.Multipart] });
     }
 
-    // Implicit body parameter would have usage flag: UsageFlags.Spread, for this case we need to do body parameter flatten
-    const bodyParameterFlatten = sdkType.kind === "model" && sdkType.usage & UsageFlags.Spread && !this.isArm();
-
-    if (schema instanceof ObjectSchema && bodyParameterFlatten) {
-      // flatten body parameter
-      const parameters = sdkHttpOperation.parameters;
-      const bodyParameter = sdkHttpOperation.bodyParam;
-      // name the schema for documentation
-      schema.language.default.name = pascalCase(op.language.default.name) + "Request";
-
-      if (!parameter.language.default.name) {
-        // name the parameter for documentation
-        parameter.language.default.name = "request";
-      }
+    if (op.convenienceApi) {
+      // Explicit body parameter @body or @bodyRoot would result to the existance of rawHttpOperation.parameters.body.property
+      // Implicit body parameter would result to rawHttpOperation.parameters.body.property be undefined
+      // see https://typespec.io/docs/libraries/http/cheat-sheet#data-types
+      const bodyParameterFlatten =
+        schema instanceof ObjectSchema &&
+        sdkType.kind === "model" &&
+        !rawHttpOperation.parameters.body?.property &&
+        !this.isArm();
+
+      if (schema instanceof ObjectSchema && bodyParameterFlatten) {
+        // flatten body parameter
+        const parameters = sdkHttpOperation.parameters;
+        const bodyParameter = sdkHttpOperation.bodyParam;
+
+        if (!parameter.language.default.name) {
+          // name the parameter for documentation
+          parameter.language.default.name = "request";
+        }
 
-      if (operationIsJsonMergePatch(sdkHttpOperation)) {
-        // skip model flatten, if "application/merge-patch+json"
-        schema.language.default.name = pascalCase(op.language.default.name) + "PatchRequest";
-        return;
-      }
```

Comment:
This change does not relate to the issue.

But we should use name from TCGC in almost all cases.